### DISCUSSION TITLE
创建文件遇到的问题

### DISCUSSION THREAD

**Discussion by rose8928** [opensumi/core#2049] [Q&A]
我创建文件不是通过FILE_COMMANDS.NEW_FILE的方式，是自定义使用fileservice.createfile直接创建，创建文件正确，但是打开的时候报错，如下图，我是自己注册组件对应的文件格式ext。
![#号报错](https://user-images.githubusercontent.com/35814595/205806263-657414ee-ef66-458f-912a-d9bc8c645f77.png)
![创建文件正确](https://user-images.githubusercontent.com/35814595/205806283-3f32078a-a3e5-406b-8b6d-580e1f5d3edf.png)

**Comment by erha19** [comment]
@rose8928 看起来是打开文件的时候路径存在问题，这里的路径不应该是  `/c:/....` 最终传递到 Node 层处理的文件路径应该为 `C:\...`，你可以本地调试看一下正常读写的文件路径是啥判断一下哪里写错了，或提供一个最小可复现的案例供我们进行问题复现。

**Comment by rose8928** [comment]
不好意思，刚才少说了一些，带#号就有问题，不带#号是可以的。
fromUri可以传递当前目录或者目录中的文件的路径，name可以传递“name.device”，content可以传递一个json（我刚测试了，.device是我注册的文档类型，如果不是注册的文档类型，是没问题的）
 async _create_file(fromUri: URI, name: string, content: object) {
    try {
      let path = fromUri.path;
      // console.log('fromUir:', fromUri);
      let stat = await this._fileService.getFileStat(fromUri.toString());
      console.log('stat:', stat);
      if (stat && !stat.isDirectory) {
        path = URI.parse(stat.uri).path.dir;
      }
      path = path.join(name);
      console.log('path:', path);
      let filePath = path.toString();
      console.log('filePath:', filePath);
      //const fileUri = URI.parse(filePath);
      const fileUri = URI.file(filePath);
      filePath = fileUri.toString();
      //const fileUri1 = URI(filePath);
      //console.log('fileUri1parse',fileUri1)
      console.log('fileUrifile:', fileUri);
      if (!isValidBasename(name)) {
        this._messageService.error(formatLocalize('validate.tree.invalidFileNameError', name));
        return;
      }
      if (await this._fileService.access(filePath)) {
        this._messageService.error(formatLocalize('validate.tree.fileNameExistsError', name));
        return;
      }

      await this._fileService.createFile(filePath);
      stat = await this._fileService.getFileStat(filePath);
      console.log('stat2:',stat);
      if (stat && content) {
        await this._fileService.setContent(stat, JSON.stringify(content));
      }
      this._commandService.executeCommand(EDITOR_COMMANDS.OPEN_RESOURCE.id, fileUri);
      return stat;
    } catch (error) {
      console.error(error);
    }
    return null;
  }



注册代码如下：

ET_DEVICE_EXT = ’.device‘
    registry.registerEditorComponent({
      component: EtDeviceEditorComponent,
      uid: editor_id,
      scheme: Schemes.file,
    });

    // 关联文件扩展名
    registry.registerEditorComponentResolver(Schemes.file, (resource, results, resolver) => {
      if (resource.uri.path.ext === ET_DEVICE_EXT) {
        return resolver([
          {
            type: 'component',
            componentId: editor_id,
            saveResource: (resource: IResource, reason: SaveReason) =>
              this.mainFrontService.saveDocument(resource, reason),
            revertResource: (resource: IResource) => this.mainFrontService.revertDocument(resource),
            undo: (resource: IResource) => this.mainFrontService.undoDocument(resource),
            redo: (resource: IResource) => this.mainFrontService.redoDocument(resource),
          },
        ]);
      }
    });

**Comment by erha19** [comment]
@rose8928  执行下面这句时，追加输出一下 URI 信息
```ts
this._commandService.executeCommand(EDITOR_COMMANDS.OPEN_RESOURCE.id, fileUri);
console.log('openuri => ', fileUri, fileUri.toString());
```
同时提供一下运行后的控制台报错和日志信息，另外，本地带 `#` 的文件确认创建成功了吧？

**Comment by rose8928** [comment]
本地以#号结尾的文件确认创建成功，本地创建test#.device，报错如下：（日志信息是vscode终端的信息吗，并没有报错，只有浏览器报错，贴图如下）感觉是URI.parse（）之后，把#号后面的当作fragment给分割开了，但是找不到具体代码在哪里。而且不是自己注册的组件没有这个问题。
![报错信息1](https://user-images.githubusercontent.com/35814595/205841424-6db4f644-296b-460a-9305-2a5d1d0769ba.png)
![错误信息3](https://user-images.githubusercontent.com/35814595/205841430-2658de44-650e-4a69-bae5-b33f86231d44.png)
![报错信息2](https://user-images.githubusercontent.com/35814595/205841455-f6bd135c-83d8-41da-86eb-f39ed71e28a2.png)
![控制台错误1](https://user-images.githubusercontent.com/35814595/205841486-0faccceb-cd22-4cca-804e-56055aa4a55e.png)
![控制台错误2](https://user-images.githubusercontent.com/35814595/205841502-803b5688-cd34-42d8-b335-d0d4a56d4b3c.png)

**Reply by erha19** [reply]
@rose8928 你这里的打开路径看着没啥问题，具体报错位置堆栈一层层往外调试看一下吧，看一下是哪里将路径处理没了，文件都是未压缩的，应该好调试的，直接页面断点就行了

**Comment by rose8928** [comment]
好的，感谢。我在代码中使用fileservice.readfile(uri:string)，发现opensumi源码中有个converturi方法，里面使用了new URI();，把类似于这种/c:/Users/huhao/Desktop/et_project/h/sdkjfjdsk#.device分割了#号后面的内容为fragment,因此发生这种错误，想问一下，如何传递feadfile里的参数可以避免此问题。

**Reply by erha19** [reply]
这里文件路径被分隔应该也是不要紧的，最终在 `uri.toString()` 下应该还是会拼接在一起的，你看一下你自己实现的 `readFile`，如果采用的是默认文件读写，可以看一下这里代码调试一下看看 https://github.com/opensumi/core/blob/9f3b00565b3e3ef3e3986062ff2dda30a25b7936/packages/file-service/src/node/disk-file-system.provider.ts#L188
